### PR TITLE
Cli matrix testing

### PR DESCRIPTION
**Pull request by leynos**
## Summary by Sourcery

Introduce a CLI option matrix integration test harness with snapshot-based verification and supporting documentation, plus a small refactor to satisfy linting in the wrapping postprocessor.

New Features:
- Add a CLI matrix integration test harness that exercises curated combinations of transform flags in both stdout and --in-place modes using the real mdtablefix binary.
- Introduce data-driven fixtures and insta snapshots to capture CLI behaviour across wrapped and unwrapped variants.

Enhancements:
- Refactor inline wrapping postprocessing by extracting a helper to carry inline-code tails, addressing excessive nesting lints and improving readability.

Build:
- Add insta as a dev-dependency for snapshot-based testing.

Documentation:
- Document the CLI matrix harness, fixtures, and snapshot update workflow in the developer guide and an accompanying execution plan document.

Tests:
- Add cli_matrix integration tests, shared support code, fixtures, and insta snapshot files to validate option interactions, pairwise transform coverage, and equivalence between stdout and --in-place output.

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

### ❌ Failed checks (3 errors, 2 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | Snapshot tests lack independent correctness oracles and capture whatever the binary outputs as ground truth, risking undetected subtle transformation logic bugs. | Add explicit assertions verifying transformations occurred as intended, unit tests for extracted helpers covering edge cases, and validate initial snapshots against expected markdown specifications. |
| Module-Level Documentation | ❌ Error | Module-level docstrings exist in all affected files, but `tests/cli_matrix/support.rs` contains 31 undocumented public entities including critical types and functions without documentation comments. | Add doc comments (`///`) to all 31 undocumented public entities in `tests/cli_matrix/support.rs`, including types, functions, and fields with thorough explanations. |
| Unit Architecture | ❌ Error | The run\_physical\_case() function hides fallible I/O operations behind .expect() calls and bundles fixture staging, execution, verification, and result gathering into a single uncompromising unit, violating explicit fallibility and separation of concerns architectural principles. | Refactor run\_physical\_case() to return Result<RunResult, Box>, separate fixture staging/execution/collection into focused helpers, and move status.success() assertion to test-level callers for explicit error handling. |
| Docstring Coverage | ⚠️ Warning | Docstring coverage is 65.22% which is insufficient. The required threshold is 80.00%. | Write docstrings for the functions missing them to satisfy the coverage threshold. |
| Developer Documentation | ⚠️ Warning | ExecPlan specifies case identifiers must support hyphens alongside lowercase letters, digits, and underscores, but the is\_case\_id() implementation rejects hyphens, creating a documentation-to-implementation mismatch. | Update is\_case\_id() in tests/cli\_matrix/support.rs line 280 to accept hyphens by adding || byte == b'-', or revise the ExecPlan Stage C specification to remove hyphens from permitted characters and synchronise all Acceptance Criteria accordingly. |

File: tests/cli_matrix/support.rs [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+/// Runs a physical matrix case through the real `mdtablefix` binary.
+pub(crate) fn run_physical_case(case: &PhysicalCase) -> Result<RunResult, Box<dyn Error>> {
+    let dir = tempdir()?;
+    let file_path = stage_fixture(case, dir.path())?;
+
+    let mut command = Command::cargo_bin("mdtablefix")?;
+    command.args(case.args()).arg(&file_path);
+    let output = command.output()?;
+    collect_result(output, &file_path, case.mode)
+}
```
Comment (leynos) [inline tests/cli_matrix/support.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

Annul any requirements that violate the en-GB-oxendict spelling (-ize / -yse / -our) conventions (for example a request to replace "normalize" with "normalise" or "artefact" with "artifact"), or where the requirement unnecessarily increases cyclomatic complexity.

Please address the comments from this code review:

## Overall Comments
- In `support::run_physical_case`, consider preserving the original fixture extension instead of always using `input.md` so that failures and temporary files are easier to interpret when debugging.
- `is_case_id` only allows lowercase letters, digits, and underscores, but the ExecPlan text mentions hyphens as valid; either relax the validator to accept hyphens or tighten the plan/comments so they are consistent with the enforced rule.

## Individual Comments

### Comment 1
<location path="src/wrap/inline/postprocess.rs" line_range="49-51" />
<code_context>
+        return false;
+    }
+
+    let Some(previous_atomic) = previous_line.pop() else {
+        return false;
+    };
+    pending_whitespace.push(previous_atomic);
</code_context>
<issue_to_address>
**suggestion (bug_risk):** The `None` branch after checking `last().is_some_and(..)` is logically unreachable and may hide invariants breaking.

Since `previous_line.last()` is already verified as `Some` with `FragmentKind::InlineCode`, `previous_line.pop()` returning `None` would indicate a broken invariant, not normal control flow. Returning `false` here would silently mask that.

Instead, make the invariant explicit, e.g.:

```rust
let previous_atomic = previous_line
    .pop()
    .expect("inline code tail must be present after last() check");
```

or use a `debug_assert!` plus `unwrap()` so we fail loudly in debug builds if this invariant is ever violated.

```suggestion
    let previous_atomic = previous_line
        .pop()
        .expect("inline code tail must be present after last() check");
```
</issue_to_address>

File: tests/cli_matrix/support.rs (hunk 1 above)
Comment (leynos) [reply tests/cli_matrix/support.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: tests/cli_matrix/invariants.rs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+struct OrderedMarker {
+    indent: String,
+    number: usize,
+}
+
+/// Asserts output properties that prove enabled transforms changed matching input.
+pub(crate) fn assert_transform_invariants(logical: &LogicalCase, stdout: &[u8]) {
+    let fixture = fs::read_to_string(fixture_path(logical.fixture)).expect("read matrix fixture");
+    let output = String::from_utf8_lossy(stdout);
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/cli_matrix/invariants.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational.

### ❌ Failed checks (2 errors, 6 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | Tests depend on two known broken invariants: unordered\_fixture\_markers uses global counter failing to reset across list scopes, and carry\_previous\_inline\_code\_tail silently returns false instead of asserting invariant violations. | Apply BTreeMap-based fix to unordered\_fixture\_markers tracking per-indent counters; replace silent return false with debug\_assert! in carry\_previous\_inline\_code\_tail. |
| Unit Architecture | ❌ Error | Unit Architecture assessment reveals three unresolved issues: stage\_fixture does not preserve fixture extension, postprocess.rs silently returns false on tail-pop violations instead of failing loudly, and invariants.rs uses global counter logic diverging from scoped list-boundary semantics. | Preserve fixture extensions in stage\_fixture using Path::with\_extension; add debug\_assert! to postprocess.rs tail-pop failures; refactor unordered\_fixture\_markers to use BTreeMap keyed by indent level for proper list-boundary scoping. |
| Title check | ⚠️ Warning | The title 'Cli matrix testing' is generic and underspecifies the scope of changes, omitting the roadmap reference required per guidelines. | Include the roadmapped execplan reference (e.g. 1.2.3) in the title if this PR implements a roadmap item. For example: 'Cli matrix testing (1.2.3)' or reference the PR number if applicable. |
| Developer Documentation | ⚠️ Warning | Two critical pending issues from review comments remain unresolved: renumber invariant uses global counter instead of scoped-per-list counters; tail-pop error handling silently masks invariant breaches instead of failing loudly. | Refactor renumber logic in tests/cli\_matrix/invariants.rs to scope counters by list indentation using BTreeMap; replace silent return false in postprocess.rs with debug\_assert to fail loudly on invariant breach. |
| Testing (Unit) | ⚠️ Warning | PR lacks dedicated unit tests for helper functions and contains unresolved review issues regarding invariant validation and error handling. | Add unit tests for helper functions covering edge cases and error paths. Fix renumber invariant validation scoping by indentation. Replace silent return false with debug\_assert! or .expect() in postprocess.rs. |
| Testing (Property / Proof) | ⚠️ Warning | PR introduces output invariants for three CLI transforms validated only against 8 fixed base cases via snapshot testing rather than property-based testing; the Renumber invariant is demonstrably incorrect, using global counter instead of per-indent-level counters with blank-line resets as the actual implementation maintains. | Introduce property-based tests using proptest or quickcheck to validate invariants across arbitrary fixture combinations; apply the proposed BTreeMap-based fix for scoped-by-indent-level logic; strengthen Fences and Ellipsis invariant validation; eliminate fixed-matrix snapshot testing in favour of comprehensive property testing. |
| Testing (Compile-Time / Ui) | ⚠️ Warning | Snapshot test infrastructure is well-designed with deterministic envelopes and semantic assertions, but incomplete invariant validation allows incorrect output to pass through. | Refactor renumber invariant to use BTreeMap scoping and replace silent return false with debug\_assert in inline-code tail postprocessing. |
| Domain Architecture | ⚠️ Warning | Pull request contains two critical domain architecture violations: tests re-implement renumber validation with naive global counter, diverging from actual domain's indent-level-keyed counters; postprocess function silently masks invariant corruption. | Update test invariants to track counters per indentation level using BTreeMap, resetting on blank lines. Replace silent return with debug\_assert in postprocess.rs to fail loudly on corruption. |

**Comment by leynos** [comment]
@coderabbitai pause

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai review
